File: G45HW2.py
import sys
import time
from pyspark import SparkContext, SparkConf
from pyspark.mllib.clustering import KMeans
import numpy as np
from computeVectorX import computeVectorX


def MRComputeStandardObjective(U, C):
    """
    This function computes the standard K-means clustering cost function on a set of points U with centers C.

    Parameters
    ----------
    U : pyspark.RDD
        The set of data points as (pos, category) where pos is a vector and category is either "A" or "B".
    C : iterable
        The centers

    Returns
    -------
    float
        The value of the cost function
    """

    return U.map(lambda x : min([sum([(xi-ci)**2 for xi,ci in zip(x[0],c)]) for c in C])).mean()


def MRComputeFairObjective(U, C):
    """
    This function computes the fair K-means clustering cost function on a set of points U with centers C.

    Parameters
    ----------
    U : pyspark.RDD
        The set of data points as (pos, category) where pos is a vector and category is either "A" or "B".
    C : iterable
        The centers

    Returns
    -------
    float
        The value of the fair cost function
    """
    DeltaA = MRComputeStandardObjective(U.filter(lambda x : x[1] == 'A'), C)
    DeltaB = MRComputeStandardObjective(U.filter(lambda x : x[1] == 'B'), C)
    return max(DeltaA, DeltaB)

def gather_partitions(pts):
    """
    pts: iterable of tuples i,x
    where i int, x point
    """
    # print(list(pts))
    if len(list(pts)) == 0:
      return []

    dim = len(list(pts)[0][1])
    K = max([p[0] for p in pts]) + 1
    cnt = [0]*K
    ans = np.zeros((K,dim))
    for i,x in pts:
        cnt[i] += 1
        ans[i] += np.array(x)
    return [(i,(cnt[i], ans[i])) for i in range(K)]

def reduce_partitions(pts):
    """
    """
    # print(list(pts))
    if len(list(pts)) == 0:
      return []

    dim = len(list(pts)[0][1])
    cnt = int(0)
    ans = [0]*dim
    for s,x in pts:
        cnt += s
        ans += x
    return [(cnt,ans)]


def MRFairLloyd(U, K, M):
    """
    Implements the Fair K-Means Clustering algorithm.

    Parameters
    ----------
    U : pyspark.RDD
        The set of data points as (pos, category) where pos is a vector and category is either "A" or "B".
    k : int
        Number of clusters
    m : int
        Number of iterations

    Returns
    -------
    list
        Final set of centroids
    """
    vectors_rdd = U.map(lambda x: x[0])
    model = KMeans.train(vectors_rdd, K, maxIterations=0)
    C = model.clusterCenters

    UA = U.filter(lambda x : x[1] == 'A').cache(); countA = UA.count()
    UB = U.filter(lambda x : x[1] == 'B').cache(); countB = UB.count()

    dim = len(C[0]) # number of dimensions of the points
    a, Ma = np.zeros(K), np.zeros((K,dim))
    b, Mb = np.zeros(K), np.zeros((K,dim))
    T = 10; gamma = 0.5

    for i in range(M):
        ret = UA.mapPartitions(lambda p : gather_partitions([(np.argmin([np.square(np.array(x[0])-c).sum() for c in C]), x[0]) for x in p])) \
                .groupByKey() \
                .mapValues(reduce_partitions) \
                .collect()
        for i,[(ai,mui)] in ret:
            a[i] = ai
            if ai > 0: Ma[i] = mui/ai
        a /= countA

        ret = UB.mapPartitions(lambda p : gather_partitions([(np.argmin([np.square(np.array(x[0])-c).sum() for c in C]),x[0]) for x in p])) \
                .groupByKey() \
                .mapValues(reduce_partitions) \
                .collect()
        for i,[(bi,mui)] in ret:
            b[i] = bi
            if bi > 0: Mb[i] = mui/bi
        b /= countB
        Ma[a == 0] = Mb[a == 0]
        Mb[b == 0] = Ma[b == 0]

        l = np.linalg.norm(Ma-Mb,axis=1)

        fixed_a = MRComputeStandardObjective(UA, Ma)/countA
        fixed_b = MRComputeStandardObjective(UB, Mb)/countB

        x = computeVectorX(fixed_a,fixed_b,a,b,l,K)

        C = [((l[i]-x[i])*Ma[i] + x[i]*Mb[i])/l[i] if l[i] > 0 else Ma[i] for i in range(K) ]
    return C


def parse_line(line):
    parts = line.strip().split(',')
    point = tuple(float(x) for x in parts[:-1])
    group = parts[-1]
    return (point, group)


def main():
    if len(sys.argv) != 5:
        print("Usage: G45HW2.py <file_path> <L> <K> <M>")
        sys.exit(1)

    file_path, L, K, M = sys.argv[1], int(sys.argv[2]), int(sys.argv[3]), int(sys.argv[4])

    # Initialize Spark context
    conf = SparkConf().setAppName("FairKMeans")
    sc = SparkContext(conf=conf)

    # Print command-line arguments
    print(f"Input file = {file_path}, L = {L}, K = {K}, M = {M}")

    # Read input points into an RDD with L partitions
    points_rdd = sc.textFile(file_path, minPartitions=L).map(parse_line).cache()

    # Print points statistics
    N = points_rdd.count()
    NA = points_rdd.filter(lambda x: x[1] == 'A').count()
    NB = points_rdd.filter(lambda x: x[1] == 'B').count()
    print(f"N = {N}, NA = {NA}, NB = {NB}")

    # Compute standard Lloyd's centroids
    start_time = time.time()
    vectors_rdd = points_rdd.map(lambda x: x[0])
    model = KMeans.train(vectors_rdd, K, maxIterations=M)
    standard_centroids = model.clusterCenters
    standard_time = int((time.time() - start_time) * 1000)  # Convert to milliseconds

    # Compute fair Lloyd's centroids
    start_time = time.time()
    fair_centroids = MRFairLloyd(points_rdd, K, M)
    fair_time = int((time.time() - start_time) * 1000)  # Convert to milliseconds

    # Compute objective functions
    start_time = time.time()
    standard_obj = MRComputeFairObjective(points_rdd, standard_centroids)
    standard_obj_time = int((time.time() - start_time) * 1000)  # Convert to milliseconds

    start_time = time.time()
    fair_obj = MRComputeFairObjective(points_rdd, fair_centroids)
    fair_obj_time = int((time.time() - start_time) * 1000)  # Convert to milliseconds

    # Print results in the exact format required
    print(f"Fair Objective with Standard Centers = {standard_obj:.4f}")
    print(f"Fair Objective with Fair Centers = {fair_obj:.4f}")
    print(f"Time to compute standard centers = {standard_time} ms")
    print(f"Time to compute fair centers = {fair_time} ms")
    print(f"Time to compute objective with standard centers = {standard_obj_time} ms")
    print(f"Time to compute objective with fair centers = {fair_obj_time} ms")

    sc.stop()


if __name__ == "__main__":
    main()

File: java/G45HW2.java
package com.example.g45hw2;

import java.util.*;
import java.util.stream.Collectors;
import java.io.Serializable;

import org.apache.spark.api.java.*;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.function.*;
import org.apache.spark.mllib.clustering.KMeans;
import org.apache.spark.mllib.clustering.KMeansModel;
import org.apache.spark.mllib.linalg.Vector;
import org.apache.spark.mllib.linalg.Vectors;
import scala.Tuple2;

public class G45HW2 {

    public static double[] computeVectorX(double fixedA, double fixedB, double[] alpha, double[] beta, double[] ell, int K) {
        double gamma = 0.5;
        double[] xDist = new double[K];
        double fA, fB;
        double power = 0.5;
        int T = 10;
        for (int t=1; t<=T; t++){
            fA = fixedA;
            fB = fixedB;
            power = power/2;
            for (int i=0; i<K; i++) {
                double temp = (1-gamma)*beta[i]*ell[i]/(gamma*alpha[i]+(1-gamma)*beta[i]);
                xDist[i]=temp;
                fA += alpha[i]*temp*temp;
                temp=(ell[i]-temp);
                fB += beta[i]*temp*temp;
            }
            if (fA == fB) {break;}
            gamma = (fA > fB) ? gamma+power : gamma-power;
        }
        return xDist;
    }

    public static class PointWithGroup implements Serializable {
        public Vector point;
        public String group;

        public PointWithGroup(Vector point, String group) {
            this.point = point;
            this.group = group;
        }
    }

    public static double computeStandardObjective(JavaRDD<PointWithGroup> U, Vector[] centers) {
        JavaRDD<Double> distances = U.map(p -> {
            double minDist = Double.MAX_VALUE;
            for (Vector c : centers) {
                double dist = squaredDistance(p.point, c);
                if (dist < minDist) minDist = dist;
            }
            return minDist;
        });
        return distances.reduce(Double::sum) / U.count();
    }

    public static double computeFairObjective(JavaRDD<PointWithGroup> U, Vector[] centers) {
        double deltaA = computeStandardObjective(U.filter(p -> p.group.equals("A")), centers);
        double deltaB = computeStandardObjective(U.filter(p -> p.group.equals("B")), centers);
        return Math.max(deltaA, deltaB);
    }

    public static Vector[] MRFairLloyd(JavaRDD<PointWithGroup> U, int K, int M) {
        JavaRDD<Vector> vectors = U.map(p -> p.point);
        KMeansModel model = KMeans.train(vectors.rdd(), K, 0); // Initial centers
        Vector[] C = model.clusterCenters();

        JavaRDD<PointWithGroup> UA = U.filter(p -> p.group.equals("A"));
        JavaRDD<PointWithGroup> UB = U.filter(p -> p.group.equals("B"));

        long countA = UA.count();
        long countB = UB.count();
        int dim = C[0].size();

        double[] a = new double[K];
        double[][] Ma = new double[K][dim];
        double[] b = new double[K];
        double[][] Mb = new double[K][dim];
        double[] l = new double[K];

        for (int iter = 0; iter < M; iter++) {
            Map<Integer, Tuple2<Double, double[]>> groupA = gatherAndReduce(UA, C, dim);
            for (int i = 0; i < K; i++) {
                Tuple2<Double, double[]> data = groupA.getOrDefault(i, new Tuple2<>(0.0, new double[dim]));
                a[i] = data._1();
                Ma[i] = scaleVector(data._2(), 1.0 / Math.max(a[i], 1e-9));
            }
            a = normalize(a, countA);

            Map<Integer, Tuple2<Double, double[]>> groupB = gatherAndReduce(UB, C, dim);
            for (int i = 0; i < K; i++) {
                Tuple2<Double, double[]> data = groupB.getOrDefault(i, new Tuple2<>(0.0, new double[dim]));
                b[i] = data._1();
                Mb[i] = scaleVector(data._2(), 1.0 / Math.max(b[i], 1e-9));
            }
            b = normalize(b, countB);

            for (int i = 0; i < K; i++) {
                if (a[i] == 0) Ma[i] = Mb[i];
                if (b[i] == 0) Mb[i] = Ma[i];
                l[i] = norm(diff(Ma[i], Mb[i]));
            }

            double fixedA = computeStandardObjective(UA, toVectors(Ma)) / countA;
            double fixedB = computeStandardObjective(UB, toVectors(Mb)) / countB;

            double[] x = computeVectorX(fixedA, fixedB, a, b, l, K);

            for (int i = 0; i < K; i++) {
                C[i] = Vectors.dense(
                        l[i] > 0
                        ? blend(Ma[i], Mb[i], x[i], l[i])
                        : Ma[i]
                );
            }
        }

        return C;
    }

    public static void main(String[] args) {
        if (args.length != 4) {
            System.err.println("Usage: FairKMeans <file_path> <L> <K> <M>");
            System.exit(1);
        }

        String filePath = args[0];
        int L = Integer.parseInt(args[1]);
        int K = Integer.parseInt(args[2]);
        int M = Integer.parseInt(args[3]);

        SparkConf conf = new SparkConf().setAppName("FairKMeans");
        JavaSparkContext sc = new JavaSparkContext(conf);

        JavaRDD<String> lines = sc.textFile(filePath, L);
        JavaRDD<PointWithGroup> data = lines.map(line -> {
            String[] parts = line.split(",");
            double[] coords = Arrays.stream(parts, 0, parts.length - 1).mapToDouble(Double::parseDouble).toArray();
            String group = parts[parts.length - 1];
            return new PointWithGroup(Vectors.dense(coords), group);
        }).cache();

        long N = data.count();
        long NA = data.filter(p -> p.group.equals("A")).count();
        long NB = data.filter(p -> p.group.equals("B")).count();
        System.out.printf("N = %d, NA = %d, NB = %d\n", N, NA, NB);

        long t0 = System.currentTimeMillis();
        KMeansModel standardModel = KMeans.train(data.map(p -> p.point).rdd(), K, M);
        Vector[] standardCenters = standardModel.clusterCenters();
        long t1 = System.currentTimeMillis();

        Vector[] fairCenters = MRFairLloyd(data, K, M);
        long t2 = System.currentTimeMillis();

        double stdObj = computeFairObjective(data, standardCenters);
        long t3 = System.currentTimeMillis();

        double fairObj = computeFairObjective(data, fairCenters);
        long t4 = System.currentTimeMillis();

        System.out.printf("Fair Objective with Standard Centers = %.4f\n", stdObj);
        System.out.printf("Fair Objective with Fair Centers = %.4f\n", fairObj);
        System.out.printf("Time to compute standard centers = %d ms\n", (t1 - t0));
        System.out.printf("Time to compute fair centers = %d ms\n", (t2 - t1));
        System.out.printf("Time to compute objective with standard centers = %d ms\n", (t3 - t2));
        System.out.printf("Time to compute objective with fair centers = %d ms\n", (t4 - t3));

        sc.stop();
    }

    // ---- Utility Methods ----

    public static double squaredDistance(Vector v1, Vector v2) {
        double sum = 0;
        for (int i = 0; i < v1.size(); i++) {
            double diff = v1.apply(i) - v2.apply(i);
            sum += diff * diff;
        }
        return sum;
    }

    public static Map<Integer, Tuple2<Double, double[]>> gatherAndReduce(JavaRDD<PointWithGroup> rdd, Vector[] centers, int dim) {
        return rdd.mapToPair(p -> {
            int best = -1;
            double bestDist = Double.MAX_VALUE;
            for (int i = 0; i < centers.length; i++) {
                double dist = squaredDistance(p.point, centers[i]);
                if (dist < bestDist) {
                    bestDist = dist;
                    best = i;
                }
            }
            return new Tuple2<>(best, p.point.toArray());
        }).groupByKey().mapValues(pts -> {
            int count = 0;
            double[] sum = new double[dim];
            for (double[] pt : pts) {
                for (int j = 0; j < dim; j++) sum[j] += pt[j];
                count++;
            }
            return new Tuple2<>((double) count, sum);
        }).collectAsMap();
    }

    public static double[] scaleVector(double[] vec, double scale) {
        return Arrays.stream(vec).map(x -> x * scale).toArray();
    }

    public static double[] diff(double[] a, double[] b) {
        double[] res = new double[a.length];
        for (int i = 0; i < a.length; i++) res[i] = a[i] - b[i];
        return res;
    }

    public static double norm(double[] v) {
        return Math.sqrt(Arrays.stream(v).map(x -> x * x).sum());
    }

    public static double[] normalize(double[] arr, long total) {
        return Arrays.stream(arr).map(x -> x / total).toArray();
    }

    public static double[] blend(double[] a, double[] b, double x, double l) {
        double[] result = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            result[i] = ((l - x) * a[i] + x * b[i]) / l;
        }
        return result;
    }

    public static Vector[] toVectors(double[][] arr) {
        return Arrays.stream(arr).map(Vectors::dense).toArray(Vector[]::new);
    }

    // computeVectorX must be implemented elsewhere and available statically
}
